azure-quantum/azure/quantum/job/session.py

```diff
@@ -0,0 +1,165 @@
+                                             **kwargs)
+
+        super().__init__(
+            workspace=workspace,
+            details=session_details,
+            **kwargs
+        )
+
+    @property
+    def details(self) -> SessionDetails:
+        return self._details
+
+    @details.setter
+    def details(self, value: SessionDetails):
+        self._details = value
+
+    @property
+    def target(self) -> "Target":
+        return self._target
+
+    def start(self) -> "Session":
+        self.workspace.start_session(self)
+        return self
+
+    def end(self) -> "Session":
+        self.workspace.end_session(self)
+        return self
+
+    def refresh(self) -> "Session":
+        self.workspace.refresh_session(self)
+        return self
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        session = self.end()
+        if isinstance(value, Exception):
+            raise
+
+
+class AlreadyHasASessionError(Exception):
+    """Exception raised when trying to start a new session `target.start_session()`
+       and the current target instance already has a session associated with it.
+    """
+
+    def __init__(self, session_id):
+        self.message = f"""The current target instance already has a session ({session_id}) associated with it.
+                           If you want to start a new session, you should obtain a new target instance.
+                           A new target instance can be obtained with `workspace.get_targets("provider_id.target_name")`
+                           Qiskit: `target` is the same concept as a Qiskit `backend`. It can be obtained with `provider.get_backend("provider_id.target_name")`.
+                           Cirq: a new target instance can be obtained with `service.get_target("provider_id.target_name")`.
+                           """
+        super().__init__(self.message)
+
+
+class SessionHost(Protocol):
+    _current_session: Optional[Session] = None
+
+    @property
+    def current_session(self) -> Optional[Session]:
+        return self._current_session
+
+    @current_session.setter
+    def current_session(self, session: Optional[Session]):
+        self._current_session = session
+
+    def get_current_session_id(self) -> Optional[str]:
+        return self.current_session.id if self.current_session else None
+
+    @abstractmethod
+    def _get_azure_workspace(self) -> "Workspace":
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_azure_target_id(self) -> str:
+        raise NotImplementedError
+
+    def start_session(
+        self,
+        session_details: Optional[SessionDetails] = None,
+        session_id: Optional[str] = None,
+        session_name: Optional[str] = None,
+        job_failure_policy: Union[str, SessionJobFailurePolicy, None] = None,
+        **kwargs
+    ) -> Session:
+        if self.current_session:
```

Comment:
One thing that we could do is to check if current_session has been ended. If it has, we can create a new session. This is nice because we could have customer's wanting to do which seems like a reasonable expectation-

backend = provider.get_backend("...")
with backend.start_session() as session:
    backend.run(circuit1)
    backend.run(circuit2)

// Do some classical logic

with backend.start_session() as session:
    backend.run(circuit3)
    backend.run(circuit4)

Reply:
Instead of persisting the current session in the target, what do you think of not persisting it at all?

The alternative would be:
```python
with backend.start_session() as session:
     session.run(circuit3)
     session.run(circuit4)
```

That would reduce the complexity of managing the session state for a given target.

We can make it in a way that the session object returned by `target.start_session()` inherits from and behaves as the target object, with the difference that there is a session associated with it.

Reply:
Since customers are used to doing backend.run() today and one of the top goals we have had is to require minimum code changes to existing code, I think session.run is not a feasible option.

Reply:
Makes sense to reduce customer friction. I'll leave as-is.